File: leetcode/src/01498_number_of_subsequences_that_satisfy_the_given_sum_condition/solution.rs
use std::iter;

struct Solution;

impl Solution {
    pub fn num_subseq(nums: Vec<i32>, target: i32) -> i32 {
        const M: i32 = 1_000_000_007;
        let mut nums = nums;
        nums.sort_unstable();
        let pows = iter::once(1)
            .chain((1..nums.len()).scan(1, |acc, _| {
                *acc = *acc * 2 % M;
                Some(*acc)
            }))
            .collect::<Vec<_>>();
        let mut res = 0;
        for (i, num) in nums.iter().enumerate() {
            let k = target - num;
            let j = nums.partition_point(|&x| x <= k);
            if j == 0 {
                continue;
            }
            if i < j {
                res = (res + pows[j - 1 - i]) % M;
            }
        }
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let nums = vec![3, 5, 6, 7];
        let target = 9;
        assert_eq!(Solution::num_subseq(nums, target), 4);

        let nums = vec![3, 3, 6, 8];
        let target = 10;
        assert_eq!(Solution::num_subseq(nums, target), 6);

        let nums = vec![2, 3, 3, 4, 6, 7];
        let target = 12;
        assert_eq!(Solution::num_subseq(nums, target), 61);
    }
}

File: leetcode/src/01498_number_of_subsequences_that_satisfy_the_given_sum_condition/solution.py
from typing import List
from itertools import accumulate, repeat
from bisect import bisect_right


class Solution:
    def numSubseq(self, nums: List[int], target: int) -> int:
        M = 10**9 + 7
        nums.sort()
        n = len(nums)
        pows = list(
            accumulate(repeat(0, n - 1), lambda acc, _: (acc * 2) % M, initial=1)
        )

        res = 0
        for i, num in enumerate(nums):
            k = target - num
            j = bisect_right(nums, k)
            if i < j:
                res = (res + pows[j - 1 - i]) % M
        return res
